### PR TITLE
[DROOLS-6455] Unify KiePMMLOutput management

### PR DESCRIPTION
**Pull request by gitgabrio**
@danielezonca @kostola  @jiripetrlik**Thank you for submitting this pull request**

See https://issues.redhat.com/browse/DROOLS-6455

Scope of this refactoring is to use Expressions inside OutputField as in other transformations. Also, to use KiePMMLOutputField in the same way in all models.

### CODE REVIEW

**Comment by gitgabrio** [comment]
Jenkins run fdb

**Comment by gitgabrio** [comment]
Jenkins run cdb

**Comment by gitgabrio** [comment]
Jenkins run fdb

**Comment by gitgabrio** [comment]
Jenkins run fdb

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/codegenfactories/KiePMMLModelFactoryUtils.java [hunk 1]
```diff
@@ -115,40 +125,65 @@ public static void setKiePMMLModelConstructor(final String generatedClassName,
...
+//            if (outputField.getResultFeature() != null) {
+//                expressions =
+//                        NodeList.nodeList(new NameExpr(RESULT_FEATURE.class.getName() + "." + outputField.getResultFeature().toString()));
+//                builder = new MethodCallExpr(builder, "withResultFeature", expressions);
+//            }
+//            Expression newOutputField = new MethodCallExpr(builder, "build");
+//            expressions = NodeList.nodeList(newOutputField);
+//            body.addStatement(new MethodCallExpr(new NameExpr("kiePMMLOutputFields"), "add", expressions));
+//        }
+//    }
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/codegenfactories/KiePMMLModelFactoryUtils.java]: To be removed

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/implementations/KiePMMLModelRetriever.java [hunk 2]
```diff
@@ -77,7 +77,7 @@ private KiePMMLModelRetriever() {
                                                                       hasClassloader))
                 .map(kiePMMLModel -> getPopulatedWithPMMLModelFields(kiePMMLModel, dataDictionary,
                                                                      model.getMiningSchema(), model.getOutput()))
-                .map(kiePMMLModel -> getPopulatedWithKiePMMLOutputFields(kiePMMLModel, model.getOutput()))
+//                .map(kiePMMLModel -> getPopulatedWithKiePMMLOutputFields(kiePMMLModel, model.getOutput()))
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/implementations/KiePMMLModelRetriever.java]: To be removed

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/implementations/KiePMMLModelRetriever.java [hunk 3]
```diff
@@ -171,14 +171,14 @@ static KiePMMLModel getPopulatedWithPMMLModelFields(final KiePMMLModel toPopulat
...
-        return toPopulate;
-    }
+//    static KiePMMLModel getPopulatedWithKiePMMLOutputFields(final KiePMMLModel toPopulate,
+//                                                            final Output output) {
+//        if (output != null) {
+//            final List<KiePMMLOutputField> converted = ModelUtils.convertToKiePMMLOutputFieldList(output);
+//            toPopulate.setKiePMMLOutputFields(converted);
+//        }
+//        return toPopulate;
+//    }
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/implementations/KiePMMLModelRetriever.java]: To be removed

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/utils/ModelUtils.java [hunk 4]
```diff
@@ -402,97 +392,97 @@ public static DATA_TYPE getDataType(DataDictionary dataDictionary, String target
...
+//        final RESULT_FEATURE resultFeature = toConvert.getResultFeature() != null ?
+//                RESULT_FEATURE.byName(toConvert.getResultFeature().value()) : null;
+//        final KiePMMLExpression kiePMMLExpression = convertToKiePMMLExpression(toConvert.getExpression());
+//        final KiePMMLOutputField.Builder builder = KiePMMLOutputField.builder(name, Collections.emptyList())
+//                .withTargetField(targetField)
+//                .withResultFeature(resultFeature)
+//                .withRank(toConvert.getRank())
+//                .withKiePMMLExpression(kiePMMLExpression);
+//        return builder.build();
+//    }
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/utils/ModelUtils.java]: To be removed (same for the rest of this class)

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/test/java/org/kie/pmml/compiler/commons/codegenfactories/KiePMMLModelFactoryUtilsTest.java [hunk 5]
```diff
@@ -154,28 +154,28 @@ public void setKiePMMLModelConstructor() {
...
+//                                                               "add");
+//        for (KiePMMLOutputField outputField : outputFields) {
+//            assertTrue(retrieved.stream()
+//                               .filter(methodCallExpr -> methodCallExpr.getArguments().size() == 1)
+//                               .map(methodCallExpr -> methodCallExpr.getArgument(0))
+//                               .filter(Expression::isMethodCallExpr)
+//                               .map(expressionArgument -> (MethodCallExpr) expressionArgument)
+//                               .anyMatch(methodCallExpr -> evaluateKiePMMLOutputFieldPopulation(methodCallExpr, outputField)));
+//        }
+//    }
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/test/java/org/kie/pmml/compiler/commons/codegenfactories/KiePMMLModelFactoryUtilsTest.java]: To be removed

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/main/java/org/kie/pmml/models/drools/ast/KiePMMLDroolsRule.java [hunk 6]
```diff
@@ -273,9 +273,9 @@ public String toString() {
         /**
          * @param name
          * @param statusToSet
-         * @param outputFields
+//         * @param outputFields
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/main/java/org/kie/pmml/models/drools/ast/KiePMMLDroolsRule.java]: To be updated

**Comment by gitgabrio** [comment]
Jenkins retest this

**Comment by gitgabrio** [comment]
Jenkins run cdb

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-regression/kie-pmml-models-regression-compiler/src/test/java/org/kie/pmml/models/regression/compiler/factories/KiePMMLRegressionTableRegressionFactoryTest.java [hunk 7]
```diff
@@ -123,35 +123,35 @@ public void setConstructor() {
...
+//        assertEquals("put", methodCallExpr.getName().asString());
+//        NodeList<Expression> arguments = methodCallExpr.getArguments();
+//        assertEquals(2, arguments.size());
+//        assertTrue(arguments.get(0) instanceof StringLiteralExpr);
+//        StringLiteralExpr stringLiteralExpr = (StringLiteralExpr)arguments.get(0);
+//        assertEquals(predictedOutputField.getName(), stringLiteralExpr.asString());
+//        assertTrue(arguments.get(1) instanceof NameExpr);
+//        NameExpr nameExpr = (NameExpr)arguments.get(1);
+//        assertEquals("result", nameExpr.getNameAsString());
+//    }
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-models/kie-pmml-models-regression/kie-pmml-models-regression-compiler/src/test/java/org/kie/pmml/models/regression/compiler/factories/KiePMMLRegressionTableRegressionFactoryTest.java]: To be removed

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-regression/kie-pmml-models-regression-compiler/src/test/java/org/kie/pmml/models/regression/compiler/factories/KiePMMLRegressionTableClassificationFactoryTest.java [hunk 8]
```diff
@@ -175,43 +175,43 @@ public void addMapPopulation() {
...
+//                                                                   StringLiteralExpr stringLiteralExpr =
+//                                                                           (StringLiteralExpr) methodCallExpr.getArguments().get(0);
+//                                                                   MethodCallExpr nestedMethodCallExpr =
+//                                                                           (MethodCallExpr) methodCallExpr.getArguments().get(1);
+//                                                                   StringLiteralExpr nestedStringLiteralExpr =
+//                                                                           (StringLiteralExpr) nestedMethodCallExpr.getArguments().get(0);
+//                                                                   return outputField.getName().equals(stringLiteralExpr.getValue())
+//                                                                           && outputField.getTargetField().get().equals(nestedStringLiteralExpr.getValue());
+//                                                               })));
+//    }
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-models/kie-pmml-models-regression/kie-pmml-models-regression-compiler/src/test/java/org/kie/pmml/models/regression/compiler/factories/KiePMMLRegressionTableClassificationFactoryTest.java]: To be removed

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-scorecard/kie-pmml-models-drools-scorecard-compiler/src/main/java/org/kie/pmml/models/drools/scorecard/compiler/factories/KiePMMLScorecardModelASTFactory.java [hunk 9]
```diff
@@ -31,7 +33,7 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import static org.kie.pmml.compiler.commons.factories.KiePMMLOutputFieldFactory.getOutputFields;
+//import static org.kie.pmml.compiler.commons.codegenfactories.KiePMMLOutputFieldFactory.getOutputFields;
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-scorecard/kie-pmml-models-drools-scorecard-compiler/src/main/java/org/kie/pmml/models/drools/scorecard/compiler/factories/KiePMMLScorecardModelASTFactory.java]: To be removed

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/main/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtils.java [hunk 10]
```diff
@@ -46,9 +47,9 @@
 
 import static org.kie.pmml.commons.Constants.MISSING_DEFAULT_CONSTRUCTOR;
 import static org.kie.pmml.commons.utils.KiePMMLModelUtils.getSanitizedClassName;
-import static org.kie.pmml.compiler.commons.factories.KiePMMLOutputFieldFactory.getOutputFields;
+//import static org.kie.pmml.compiler.commons.codegenfactories.KiePMMLOutputFieldFactory.getOutputFields;
 import static org.kie.pmml.compiler.commons.utils.JavaParserUtils.MAIN_CLASS_NOT_FOUND;
-import static org.kie.pmml.compiler.commons.codegenfactories.KiePMMLModelFactoryUtils.addKiePMMLOutputFieldsPopulation;
+//import static org.kie.pmml.compiler.commons.codegenfactories.KiePMMLModelFactoryUtils.addKiePMMLOutputFieldsPopulation;
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/main/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtils.java]: To cleanup

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/main/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtils.java [hunk 11]
```diff
@@ -86,9 +87,9 @@ public static CompilationUnit getKiePMMLModelCompilationUnit(final DataDictionar
         MINING_FUNCTION miningFunction = MINING_FUNCTION.byName(model.getMiningFunction().value());
         final ConstructorDeclaration constructorDeclaration = modelTemplate.getDefaultConstructor().orElseThrow(() -> new KiePMMLInternalException(String.format(MISSING_DEFAULT_CONSTRUCTOR, modelTemplate.getName())));
         String targetField = getTargetFieldName(dataDictionary, model).orElse(null);
-        List<KiePMMLOutputField> outputFields = getOutputFields(model);
+//        List<OutputField> outputFields = model.getOutput().getOutputFields();
         setConstructor(model, constructorDeclaration, modelTemplate.getName(), targetField, miningFunction);
-        addKiePMMLOutputFieldsPopulation(constructorDeclaration.getBody(), outputFields);
+//        addKiePMMLOutputFieldsPopulation(constructorDeclaration.getBody(), outputFields);
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/main/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtils.java]: To cleanup

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/ProcessingDTO.java [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+        this.orderedReasonCodes = orderedReasonCodes;
+    }
+
+    /**
+     *
+     * @param defineFunctions
+     * @param derivedFields
+     * @param outputFields
+     * @param kiePMMLNameValues a <b>mutable</b> list of <code>KiePMMLNameValue</code>
+     * @param orderedReasonCodes a <b>mutable</b> list
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/ProcessingDTO.java:73]: What about avoid codepath like `getXYZ().add(...)` and create methods like `addKiePMMLNameValue(...)` and also always returns immutable list using the getter

Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/ProcessingDTO.java:73]: @danielezonca done

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/main/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtils.java (hunk 11 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/main/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtils.java]: @danielezonca done

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/main/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtils.java (hunk 10 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/main/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtils.java]: @danielezonca done

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-scorecard/kie-pmml-models-drools-scorecard-compiler/src/main/java/org/kie/pmml/models/drools/scorecard/compiler/factories/KiePMMLScorecardModelASTFactory.java (hunk 9 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-scorecard/kie-pmml-models-drools-scorecard-compiler/src/main/java/org/kie/pmml/models/drools/scorecard/compiler/factories/KiePMMLScorecardModelASTFactory.java]: @danielezonca done

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-regression/kie-pmml-models-regression-compiler/src/test/java/org/kie/pmml/models/regression/compiler/factories/KiePMMLRegressionTableClassificationFactoryTest.java (hunk 8 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-models/kie-pmml-models-regression/kie-pmml-models-regression-compiler/src/test/java/org/kie/pmml/models/regression/compiler/factories/KiePMMLRegressionTableClassificationFactoryTest.java]: @danielezonca done

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-regression/kie-pmml-models-regression-compiler/src/test/java/org/kie/pmml/models/regression/compiler/factories/KiePMMLRegressionTableRegressionFactoryTest.java (hunk 7 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-models/kie-pmml-models-regression/kie-pmml-models-regression-compiler/src/test/java/org/kie/pmml/models/regression/compiler/factories/KiePMMLRegressionTableRegressionFactoryTest.java]: @danielezonca done

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/main/java/org/kie/pmml/models/drools/ast/KiePMMLDroolsRule.java (hunk 6 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/main/java/org/kie/pmml/models/drools/ast/KiePMMLDroolsRule.java]: @danielezonca done

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/test/java/org/kie/pmml/compiler/commons/codegenfactories/KiePMMLModelFactoryUtilsTest.java (hunk 5 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/test/java/org/kie/pmml/compiler/commons/codegenfactories/KiePMMLModelFactoryUtilsTest.java]: @danielezonca done

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/utils/ModelUtils.java (hunk 4 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/utils/ModelUtils.java]: @danielezonca done

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/implementations/KiePMMLModelRetriever.java (hunk 3 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/implementations/KiePMMLModelRetriever.java]: @danielezonca done

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/implementations/KiePMMLModelRetriever.java (hunk 2 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/implementations/KiePMMLModelRetriever.java]: @danielezonca done

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/codegenfactories/KiePMMLModelFactoryUtils.java (hunk 1 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/codegenfactories/KiePMMLModelFactoryUtils.java]: @danielezonca done

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/KiePMMLOutputField.java [hunk 13]
```diff
@@ -68,14 +79,41 @@ public KiePMMLExpression getKiePMMLExpression() {
...
+    public Object evaluate(final ProcessingDTO processingDTO) {
+        switch (resultFeature) {
+            case PREDICTED_VALUE:
+                return evaluatePredictedValue(processingDTO);
+            case REASON_CODE:
+                return evaluateReasonCodeValue(processingDTO);
+            case TRANSFORMED_VALUE:
+                return evaluateTransformedValue(processingDTO);
+            default:
+                return null;
```
Comment (jiripetrlik) [inline kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/KiePMMLOutputField.java:91]: What about to use Optional? Maybe it is better than just use null.

File: kie-pmml-trusty/kie-pmml-commons/src/test/java/org/kie/pmml/commons/CommonTestingUtility.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+ */
+package org.kie.pmml.commons;
+
+import java.util.List;
+
+import org.kie.pmml.commons.model.KiePMMLModel;
+import org.kie.pmml.commons.model.ProcessingDTO;
+import org.kie.pmml.commons.model.tuples.KiePMMLNameValue;
+
+public class CommonTestingUtility {
```
Comment (jiripetrlik) [inline kie-pmml-trusty/kie-pmml-commons/src/test/java/org/kie/pmml/commons/CommonTestingUtility.java]: I'm not sure about this class. It works only with ProcessingDTO. Can the name be more specific?

**Review by jiripetrlik — COMMENTED** [review]
I have two comments/questions.

File: kie-pmml-trusty/kie-pmml-commons/src/test/java/org/kie/pmml/commons/CommonTestingUtility.java (hunk 14 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-commons/src/test/java/org/kie/pmml/commons/CommonTestingUtility.java]: @jiripetrlik 
Removed

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/KiePMMLOutputField.java (hunk 13 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/KiePMMLOutputField.java:91]: @jiripetrlik 
Proabably you're right, but to keep consistency (and code clarity) I should change it also in other classes (KiePMMLExpressions, KiePMMLDerivedField, KiePMMLDefineFunciont); I would prefer to do that in a specific PR

**Review by danielezonca — APPROVED** [review]

**Review by jiripetrlik — APPROVED** [review]

**Pull request merged**